### PR TITLE
UI Automation in Windows Console: rename is21H1Plus to isImprovedTextRangeAvailable

### PR DESCRIPTION
**Pull request by codeofdusk**
### Link to issue number:
microsoft/terminal#9239

### Summary of the issue:
Windows 10 21H1 has been released to insiders with significantly reduced scope (i.e. excluding the UIA changes it was originally set to include), making our internal names misleading. Additionally, there are plans to "undock" conhost from Windows, removing the guarantee that a given Windows version will run a particular version of the console (it's technically already possible to run newer console on older Windows, I do it for testing).

### Description of how this pull request fixes the issue:
Renames:

* `NVDAObjects.UIA.winConsoleUIA.is21H1Plus` -> `NVDAObjects.UIA.winConsoleUIA.isImprovedTextRangeAvailable`
* `NVDAObjects.UIA.winConsoleUIA.consoleUIATextInfo` -> `NVDAObjects.UIA.winConsoleUIA.ConsoleUIATextInfo` (Start class name with upper case)
* `NVDAObjects.UIA.winConsoleUIA.consoleUIATextInfoPre21H1` -> `NVDAObjects.UIA.winConsoleUIA.ConsoleUIATextInfoWorkaroundEndInclusive` (because much of the implementation deals with the fact that before microsoft/terminal#4018 text ranges had both end points inclusive, necessitating workarounds for `expand`, `collapse`, `compareEndPoints`, `setEndPoint`, etc)

Also moves the `text` override to the workaround `textInfo` as #10036 is no longer reproducible in new console.
### Testing performed:
Verified that console still works as intended. These names were used internally only.

### Known issues with pull request:
None.

### Change log entry:
Changes for developers:
```
- ``NVDAObjects.UIA.winConsoleUIA.is21H1Plus`` renamed ``NVDAObjects.UIA.winConsoleUIA.isImprovedTextRangeAvailable``. (#12094)
- ``NVDAObjects.UIA.winConsoleUIA.consoleUIATextInfo`` renamed to start class name with upper case. (#12094)
- ``NVDAObjects.UIA.winConsoleUIA.consoleUIATextInfoPre21H1`` renamed ``NVDAObjects.UIA.winConsoleUIA.ConsoleUIATextInfoWorkaroundEndInclusive``. (#12094)
  - The implementation works around both end points being inclusive (in text ranges) before microsoft/terminal#4018
  - Workarounds for ``expand``, ``collapse``, ``compareEndPoints``, ``setEndPoint``, etc
```

### Code Review Checklist:

This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review and confirm you have considered the following items.
Mark items you have considered by checking them.
You can do this when editing the Pull request description with an x: `[ ]` becomes `[x]`.
You can also check the checkboxes after the PR is created.

- [x] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual tests.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.

### CODE REVIEW

**Comment by codeofdusk** [comment]
Cc @feerrenrut, @michaelDCurran, @carlos-zamora, @DHowett.

File: source/NVDAObjects/UIA/winConsoleUIA.py [hunk 1]
```diff
@@ -348,7 +348,7 @@ def _get_windowThreadID(self):
 			threadID = super().windowThreadID
 		return threadID
 
-	def _get_is21H1Plus(self):
+	def _get_isPostUIARefactor(self):
```
Comment (feerrenrut) [inline source/NVDAObjects/UIA/winConsoleUIA.py:351]: This isn't very descriptive. If the main feature that consoleUIATextInfo relies on is GetVisibleRanges then perhaps it should be called 'isGetVisibleRangeAvailable`

Comment (codeofdusk) [reply source/NVDAObjects/UIA/winConsoleUIA.py:351]: This, too, is somewhat misleading I think. `GetVisibleRanges` was available in the bad UIA implementation as well, but it returned incorrect results in some cases.

**Comment by codeofdusk** [comment]
Essentially, `GetVisibleRanges` is a convenient way to check whether microsoft/terminal#4018 is available, as that work fixed many bugs in the UIA implementation.

**Comment by feerrenrut** [comment]
Is there any other way to check for that? If not, then please just change the names and docs here to be really explicit about what you are trying to determine and why.

Perhaps a name like `IsImprovedUIATextRangeAvailable` with a docstring including all the notes you have given me in comments here, and also pointing to that PR.

**Comment by codeofdusk** [comment]
> Is there any other way to check for that?

Sadly I don't think so, CC @miniksa and @carlos-zamora.

**Comment by josephsl** [comment]
Hi, one possible idea might be the ability to check for conhost.exe file metadata, as Bill pointed out that it is possible to test newer versions of this binary. I tried this approach but couldn’t get it to work (I used looking up app version for conhost.exe using file metadata method defined in app module class to no avail). Thanks.

**Comment by miniksa** [comment]
> > Is there any other way to check for that?
> 
> Sadly I don't think so, CC @miniksa and @carlos-zamora.

We are strongly advised by the Windows compatibility experts that detecting functionality is always supreme to detecting version, so we would have to say that we prefer the existing method of testing for the functionality to know you're on an appropriate version.

However, if you must go beyond that, you might be able to use [VerifyVersionInfo](https://docs.microsoft.com/windows/win32/api/winbase/nf-winbase-verifyversioninfow) of which I have a sample here https://github.com/microsoft/terminal/blob/22fd06e19b3e564d448d60ebf34ffd63764f8080/src/cascadia/TerminalSettingsModel/CascadiaSettings.cpp#L1158-L1168. I'm obviously guilty of ignoring their strong recommendation and checking the version there... but `VerifyVersionInfo` is preferred by the compat team to getting the number and doing the comparison yourself with one of the `GetVersion*` APIs.

Do note that generally speaking, the Windows compatibility team has installed a number of measures to make it very difficult for applications to get the version of the operating system and force them to feature detect instead. So you may get a "version lie" out of that API or any of the others that is crafted to the maximum version that NVDA is [pre-manifested](https://docs.microsoft.com/windows/win32/sbscs/application-manifests#supportedos) to support.

**Comment by codeofdusk** [comment]
I've updated comments to express intent more clearly, but haven't yet changed any names yet.

Something like `isImprovedTextRangeAvailable` as a replacement for `is21H1Plus` sounds good, but what should the new name for `ConsoleUIATextInfoPre21H1` be?

**Comment by codeofdusk** [comment]
@feerrenrut I settled on `TextInfoWorkaroundEndInclusive` (see description).

File: source/NVDAObjects/UIA/winConsoleUIA.py [hunk 2]
```diff
@@ -99,17 +99,17 @@ def __ne__(self, other):
...
 
 
-class consoleUIATextInfoPreRefactor(consoleUIATextInfo):
-	"""Fixes expand/collapse on end inclusive UIA text ranges, uses rangeFromPoint
-	instead of broken GetVisibleRanges for bounding, and implements word
-	movement support."""
+class TextInfoWorkaroundEndInclusive(TextInfo):
+	"""Implementation of various workarounds for pre-microsoft/terminal#4018
+	conhost: fixes expand/collapse, uses rangeFromPoint instead of broken
+	GetVisibleRanges for bounding, and implements word movement support."""
```
Comment (feerrenrut) [inline source/NVDAObjects/UIA/winConsoleUIA.py:112]: This looks good.

**Review by feerrenrut — COMMENTED** [review]
Let me know if you will change the TextInfo. Either way I think this can be merged.

File: source/NVDAObjects/UIA/winConsoleUIA.py [hunk 3]
```diff
@@ -16,7 +16,7 @@
 from ..window import Window
 
 
-class consoleUIATextInfo(UIATextInfo):
+class TextInfo(UIATextInfo):
```
[the remark this answers is not in the document]
Comment (codeofdusk) [reply source/NVDAObjects/UIA/winConsoleUIA.py:19]: Yes, but then we'll have `ConsoleUIATextInfo` and `TextInfoWorkaroundEndInclusive`: one name mentions console UIA and one doesn't.

[the remark this answers is not in the document]
Comment (feerrenrut) [reply source/NVDAObjects/UIA/winConsoleUIA.py:19]: I'm not sure I understand why that is a problem. Ideally `ConsoleUIATextInfo` would tell you something about how it is different from a `UIATextInfo` or a `TextInfo`. But at least giving it a unique name means that text tools can be used on it. 

But like I said, lots of other classes are called `TextInfo` so it's not introducing something new. It just means that smarter tools that understand namespaces must be used.

File: source/NVDAObjects/UIA/winConsoleUIA.py [hunk 4]
```diff
@@ -33,7 +33,7 @@ def __init__(self, obj, position, _rangeObj=None):
 				log.warning("Couldn't get bounding range for console", exc_info=True)
 				# Fall back to presenting the entire buffer.
 				_rangeObj, collapseToEnd = None, None
-		super(consoleUIATextInfo, self).__init__(obj, position, _rangeObj)
+		super(TextInfo, self).__init__(obj, position, _rangeObj)
```
Comment (lukaszgo1) [inline source/NVDAObjects/UIA/winConsoleUIA.py:36]: Can't you just use Python 3 `super` here? Same goes for all other modified calls to `super` below.

File: source/NVDAObjects/UIA/winConsoleUIA.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (codeofdusk) [reply source/NVDAObjects/UIA/winConsoleUIA.py:19]: Module-level internal consistency (and, as you point out, consistency with the rest of NVDA).

**Comment by codeofdusk** [comment]
Ready for a merge whenever you are.

File: source/NVDAObjects/UIA/winConsoleUIA.py [hunk 5]
```diff
@@ -90,26 +90,17 @@ def move(self, unit, direction, endPoint=None):
...
-			return ' '
-		else:
-			return res
-
 
-class consoleUIATextInfoPre21H1(consoleUIATextInfo):
-	"""Fixes expand/collapse on end inclusive UIA text ranges, uses rangeFromPoint
-	instead of broken GetVisibleRanges for bounding, and implements word
-	movement support."""
+class TextInfoWorkaroundEndInclusive(TextInfo):
```
Comment (feerrenrut) [inline source/NVDAObjects/UIA/winConsoleUIA.py:100]: The problem with using `TextInfo` as a name is that code like this becomes quite ambiguous for a reader. I think it is unnecessarily ambiguous. Consistency is worthwhile if the convention scales, I'm not sure this convention really makes anything better, other than allowing the developer a way out of coming up with a descriptive name.

Code is read many more times than it is written. Better to optimize for the reader IMO.

Comment (feerrenrut) [reply source/NVDAObjects/UIA/winConsoleUIA.py:100]: When I say ambiguous, I mean that you have to be aware that there is a class within this module called TextInfo, and check that the import doesn't do `from textInfos import TextInfo`.

Comment (codeofdusk) [reply source/NVDAObjects/UIA/winConsoleUIA.py]: OK, the name is now `ConsoleUIATextInfo` (with capitalization consistent with the rest of NVDA).

In the long term (i.e. once microsoft/terminal#6986 has been resolved), I hope to transition to just using the standard `UIATextInfo` implementation for consoles where `isImprovedTextRangeAvailable`... but that's in a follow-up PR.

**Review by feerrenrut — DISMISSED** [review]
I'll look to merge this on Monday.

**Review by feerrenrut — DISMISSED** [review]
Thanks @codeofdusk. This is being merged into Beta, it will be merged back into alpha/master shortly.

**Comment by feerrenrut** [comment]
A note @codeofdusk, this requires a change log. Code is only "internal use" if the class / function is prefixed with an underscore. Otherwise it is considered an API change and should be documented.

**Comment by codeofdusk** [comment]
Something like this in dev changes:

* `WinConsoleUIA.is21H1Plus` has been renamed to `WinConsoleUIA.isImprovedTextRangeAvailable` and `ConsoleUIATextInfoPre21H1` has been renamed to `ConsoleUIATextInfoWorkaroundEndInclusive` due to Microsoft changes
* `consoleUIATextInfo` has been renamed to `ConsoleUIATextInfo`

**Comment by feerrenrut** [comment]
I have updated the description for you.

**Comment by codeofdusk** [comment]
LGTM.

**Review by feerrenrut — APPROVED** [review]

**Pull request merged**